File: ch3/typed_tagless_final_abstract.hs
-- the (previously) untyped expression language of TAPL chapter 3,
--   using a typed tagless final representation
--   a la http://okmij.org/ftp/tagless-final/course/lecture.pdf

-- look ma, no GADTs or type families

-- this one permits multiple interpreters

import Prelude hiding (succ)

class Interp repr where
  t :: repr Bool
  f :: repr Bool
  ifThenElse :: repr Bool -> repr a -> repr a -> repr a
  z :: repr Int
  succ :: repr Int -> repr Int
  pred :: repr Int -> repr Int
  isZero :: repr Int -> repr Bool

newtype Eval a = Eval { eval :: a }

instance Interp Eval where
  t = Eval True
  f = Eval False

  ifThenElse (Eval True) t _ = t
  ifThenElse (Eval False) _ f = f

  z = Eval 0

  succ (Eval n) = Eval (1 + n)

  pred (Eval n)
    | n == 0 = Eval 0
    | otherwise = Eval (n - 1)

  isZero (Eval n) = Eval (n == 0)

newtype PPrint a = PPrint { pprint :: String }

instance Interp PPrint where
  t = PPrint "t"
  f = PPrint "f"

  ifThenElse (PPrint t1) (PPrint t2) (PPrint t3) =
    PPrint $ "ifThenElse (" ++ t1 ++ ") (" ++ t2 ++ ") (" ++ t3 ++ ")"

  z = PPrint "z"

  succ (PPrint n) = PPrint $ "succ (" ++ n ++ ")"

  pred (PPrint n) = PPrint $ "pred (" ++ n ++ ")"

  isZero (PPrint n) = PPrint $ "isZero (" ++ n ++ ")"

File: ch3/typed_tagless_final_abstract.cpp
// the typed tagless final abstract approach,
//   roughly ported to C++17

#include <iostream>
#include <string>

struct eval {
    static constexpr bool t = true;
    static constexpr bool f = false;
    
    template<class T>
    static constexpr T ifThenElse(bool cond, T t, T f) noexcept
    {
        return cond ? t : f;
    }

    static constexpr int z = 0;

    static constexpr int succ(int n) noexcept { return n + 1; }

    static constexpr int pred(int n) noexcept { return (n == 0) ? 0 : (n - 1); }

    static constexpr bool isZero(int n) noexcept { return n == 0; }
};

struct pprint {
    inline static const std::string t = "t";
    inline static const std::string f = "f";

    static std::string ifThenElse(const std::string& cond,
            const std::string& t, const std::string& f)
    {
        return "ifThenElse(" + cond + ", " + t + ", " + f + ")";
    }

    inline static const std::string z = "z";

    static std::string succ(const std::string& n)
    {
        return "succ(" + n + ")";
    }

    static std::string pred(const std::string& n)
    {
        return "pred(" + n + ")";
    }

    static std::string isZero(const std::string& n)
    {
        return "isZero(" + n + ")";
    }
};

template<class Interp> auto exampleExpr =
  Interp::ifThenElse(
    Interp::isZero(Interp::succ(Interp::z)),
    Interp::pred(Interp::succ(Interp::z)),
    Interp::succ(Interp::succ(Interp::z)));

int main()
{
    std::cout << exampleExpr<eval> << '\n';
    std::cout << exampleExpr<pprint> << '\n';
}
